### PR TITLE
feat: Integrate the block-sdk!!

### PR DESCRIPTION
**Pull request by g3ndo9kar9**
### In This PR 
- I integrate the block-sdk, that is
    - I integrate the mev, free, default lanes (ordered accordingly) into the `LaneMempool` which will be used as the app's `Mempool`
    - I integrate the Prepare / ProcessProposal handlers and custom CheckTx handler
    - Modules
        - I integrate the `x/auction` module, and the corresponding `AnteHandler` for the auction module
    - I integrate the `ante.IgnoreDecorator` with the `freeLane`,  and the `DeductFees` ante-handler: this will skip the `DeductFees` ante-handler for any txs that originate from the `freeLane`
### Testing
- We've tested this integration against our own set of load-tests. These tests, spin up a network of 5 nodes (3 val, 1 seed, 1 full). Perform various perturbations against them (killing, restarting, disconnecting from the network, etc.), and periodically send a load of 350 tx/sec for 90s. The output from one of our trials is [here](https://snapshots.raintank.io/dashboard/snapshot/y23ErIIGdQEL8AXCbFdv4grb2hxldbMz?orgId=2), happy to go through these in more depth / review the methodology.
- In order for the e2e tests to work, I had to modify the `ante_no_seq` ante-handlers to ignore signature verification, while incrementing sequence #s, I leave a comment in the section describing why.
### Let's Discuss
- I chose to use stable defaults [here](https://github.com/skip-mev/osmosis/pull/1/files#diff-4aaa0025728f832e685e0651c0febf270306c9150833063c1c9a023c02bfa14aR15) (tx-capacity of each lane) and [here](https://github.com/skip-mev/osmosis/pull/1/files#diff-4aaa0025728f832e685e0651c0febf270306c9150833063c1c9a023c02bfa14aR26) (block-space allocated to each lane), let me know if any of the parameters are confusing or should be changed
- regarding the auction-module
    -  I imagine that the full integration requires the v23 upgrade-handler (store-migration, genesis-state / parameters, etc.), I am happy to implement those as a separate PR if that is desired
    - `Auction` module's parameters are [here](https://github.com/skip-mev/block-sdk/blob/31def89bc1fa8e3460a11b53030b6d965da860f4/x/auction/types/genesis.pb.go#L76), take a look at these or we can go ahead w/ the [defaults](https://github.com/skip-mev/block-sdk/blob/31def89bc1fa8e3460a11b53030b6d965da860f4/x/auction/types/params.go#L11), some things to consider
       - What denomination would we like the auction to accept (there can only be a single denom accepted)
       - What would be the desired max bundle size (maximal # of txs that a bid can include)
       - What address would we like the auction revenue to be sent to? How much of it should we give to proposers?
- re the free-lane: We currently are matching the `distributiontypes.WithdrawDelegatorRewards` tx, the process of adding more txs to this lane is simple, we just need to identify which types we want!

### CODE REVIEW

File: app/lanes.go [hunk 1]
```diff
@@ -0,0 +1,108 @@
...
+	defaultlane "github.com/skip-mev/block-sdk/lanes/base"
+	freelane "github.com/skip-mev/block-sdk/lanes/free"
+	mevlane "github.com/skip-mev/block-sdk/lanes/mev"
+)
+
+const (
+	maxTxPerMEVLane = 500 // this is the maximum # of bids that will be held in the app-side in-memory mempool
+	maxTxPerFreeLane = 1000 // this is the maximum # of free-txs that will be held in the app-side in-memory mempool
+	maxTxPerDefaultLane = 3000 // all other txs
+)
```
Comment (davidterpay) [inline app/lanes.go:18]: good defaults

File: app/lanes.go [hunk 2]
```diff
@@ -0,0 +1,108 @@
...
+	freeLaneBlockspacePercentage = math.LegacyMustNewDecFromStr("0.05")
+	defaultLaneBlockspacePercentage = math.LegacyMustNewDecFromStr("0.85")
+	mevLaneBlockspacePercentage = math.LegacyMustNewDecFromStr("0.1")
+)
+
+// CreateLanes walks through the process of creating the lanes for the block sdk. In this function
+// we create three separate lanes - MEV, Free, and Default - and then return them.
+//
+// NOTE: Application Developers should closely replicate this function in their own application.
+func CreateLanes(app *OsmosisApp, txConfig client.TxConfig) (*mevlane.MEVLane, *base.BaseLane, *base.BaseLane) {
```
Comment (davidterpay) [inline app/lanes.go]: probably dont need the note

File: app/lanes.go [hunk 3]
```diff
@@ -0,0 +1,108 @@
...
+	// 1. Create the signer extractor. This is used to extract the expected signers from
+	// a transaction. Each lane can have a different signer extractor if needed.
+	signerAdapter := signerextraction.NewDefaultAdapter()
+
+	// 2. Create the configurations for each lane. These configurations determine how many
+	// transactions the lane can store, the maximum block space the lane can consume, and
+	// the signer extractor used to extract the expected signers from a transaction.
+	//
+	// IMPORTANT NOTE: If the block sdk module is utilized to store lanes, than the maximum
+	// block space will be replaced with what is in state / in the genesis file.
```
Comment (davidterpay) [inline app/lanes.go]: doesnt apply

File: app/lanes.go [hunk 4]
```diff
@@ -0,0 +1,108 @@
...
+
+	// Create a mev configuration that accepts 1000 transactions and consumes 20% of the
+	// block space.
```
Comment (davidterpay) [inline app/lanes.go]:
```suggestion
	// Create a mev configuration that accepts maxTxPerMEVLane transactions and consumes mevLaneBlockspacePercentage of the
	// block space.
```

File: app/lanes.go [hunk 5]
```diff
@@ -0,0 +1,108 @@
...
+		Logger:          app.Logger(),
+		TxEncoder:       txConfig.TxEncoder(),
+		TxDecoder:       txConfig.TxDecoder(),
+		MaxBlockSpace:   mevLaneBlockspacePercentage,
+		SignerExtractor: signerAdapter,
+		MaxTxs:          maxTxPerMEVLane,
+	}
+
+	// Create a free configuration that accepts 1000 transactions and consumes 5% of the
+	// block space.
```
Comment (davidterpay) [inline app/lanes.go]: same as above

File: app/lanes.go [hunk 6]
```diff
@@ -0,0 +1,108 @@
...
+		Logger:          app.Logger(),
+		TxEncoder:       txConfig.TxEncoder(),
+		TxDecoder:       txConfig.TxDecoder(),
+		MaxBlockSpace:   freeLaneBlockspacePercentage,
+		SignerExtractor: signerAdapter,
+		MaxTxs:          maxTxPerFreeLane,
+	}
+
+	// Create a default configuration that accepts 1000 transactions and consumes 60% of the
+	// block space.
```
Comment (davidterpay) [inline app/lanes.go]: same as above

File: app/app.go [hunk 7]
```diff
@@ -352,6 +360,21 @@ func NewOsmosisApp(
...
 	app.sm.RegisterStoreDecoders()
 
+	// initialize lanes + mempool
+	mevLane, defaultLane, freeLane := CreateLanes(app, txConfig)
+
+	// create the mempool
+	lanedMempool, err := block.NewLanedMempool(
+		app.Logger(),
+		[]block.Lane{mevLane, defaultLane, freeLane},
+	)
```
Comment (davidterpay) [inline app/app.go]: the default lane is going to match all transactions that are in the free lane, so i recommend we switch to mev, free, default

File: app/app.go [hunk 8]
```diff
@@ -403,6 +471,19 @@ func NewOsmosisApp(
 	return app
 }
 
+// CheckTx will check the transaction with the provided checkTxHandler. We override the default
+// handler so that we can verify bid transactions before they are inserted into the mempool.
+// With the POB CheckTx, we can verify the bid transaction and all of the bundled transactions
```
Comment (davidterpay) [inline app/app.go]:
```suggestion
// With the Block SDK CheckTx, we can verify the bid transaction and all of the bundled transactions
```

**Review by davidterpay — COMMENTED** [review]
where is the genesis set up? I dont see us configuring it in the upgrade handler

**Comment by davidterpay** [comment]
just order of lanes + upgrade handler

File: app/lanes.go [hunk 9]
```diff
@@ -0,0 +1,108 @@
...
+
+	// 3. Create the match handlers for each lane. These match handlers determine whether or not
+	// a transaction belongs in the lane.
+
+	// Create the final match handler for the mev lane.
+	factory := mevlane.NewDefaultAuctionFactory(txConfig.TxDecoder(), signerAdapter)
+	mevMatchHandler := factory.MatchHandler()
+
+	// Create the final match handler for the free lane.
+	freeMatchHandler := freelane.DefaultMatchHandler()
```
Comment (Eric-Warehime) [inline app/lanes.go:78]: The only one they were 100% sure of when we talked w/ them was claiming staking rewards--which I don't think is what the default match handler has.

File: app/ante_no_seq.go [hunk 10]
```diff
@@ -59,12 +74,21 @@ func NewAnteHandler(
...
+			deductFeeDecorator,
+			blockSDKParams.freeLane,
+		),
 		ante.NewSetPubKeyDecorator(ak), // SetPubKeyDecorator must be called before all signature verification decorators
 		ante.NewValidateSigCountDecorator(ak),
 		ante.NewSigGasConsumeDecorator(ak, sigGasConsumer),
-		ante.NewSigVerificationDecorator(ak, signModeHandler),
-		// ante.NewIncrementSequenceDecorator(ak),
+		// ante.NewSigVerificationDecorator(ak, signModeHandler) <-- removed this to prevent failures resulting from invalid tx orders in e2e
+		ante.NewIncrementSequenceDecorator(ak),
```
Comment (g3ndo9kar9) [inline app/ante_no_seq.go:85]: notice: This switch is made, as the priority nonce-mempool (underlying mempool used by the block-sdk) requires that all txs have a unique sequence number. As such, we remove the signature verification ante-handler (to ignore the sequence # check in e2e txs), but still increment the seq. to get around over-writing txs in the mempool

Comment (Eric-Warehime) [inline app/ante_no_seq.go:85]: This looks weird to me. Seems like we should end up w/ SigVerificationDecorator enabled right?

Comment (g3ndo9kar9) [reply app/ante_no_seq.go:85]: The reason we ignore sig verification is that it's possible that the sequence # of txs is incorrect, the sigVerificationDecorator is where this is checked





[3:00](https://skip-protocol.slack.com/archives/D067ERWBXR9/p1705618823070209)
here: https://github.com/cosmos/cosmos-sdk/blob/03d578b9de5136ffc658f2b0c07434373b07b7da/x/auth/ante/sigverify.go#L268

File: app/lanes.go (hunk 9 above)
Comment (g3ndo9kar9) [inline app/lanes.go]: The FreeLane match-handler matches to (staking.Delegate, Redelegate, Unbond). I wasn't sure whether it made sense to pre-emptively implement, or point out in PR and resolve in discussion, so I did not update the MatchHandler

Comment (g3ndo9kar9) [reply app/lanes.go]: Happy to update the MatchHandler accordingly tho

**Review by Eric-Warehime — APPROVED** [review]

**Pull request closed without merging**